R/data.table.R

```diff
@@ -1027,14 +1027,14 @@ replace_dot_alias = function(e) {
           lhs = jsub[[2L]]
           jsub = jsub[[3L]]
           if (is.name(lhs)) {
-            if (deparse(lhs) == '.SD') lhs = sdvars else lhs = as.character(lhs)
+            if (lhs == as.name('.SD')) lhs = sdvars else lhs = as.character(lhs)
           } else {
-            # if (deparse(lhs) == 'names(.SD)') {
-              # lhs = sdvars
-            # } else {
-              # e.g. (MyVar):= or get("MyVar"):=
-              lhs = eval(lhs, parent.frame(), parent.frame())
-            # }
+            #i.e lhs is names(.SD) || setdiff(names(.SD), cols) || (cols)
+            if (lhs[[1]] == as.name('names') && lhs[[2]] == as.name('.SD')) lhs = sdvars
+            for (i in seq_along(lhs)[-1]){
+              if (lhs[[i]] == as.name('names(.SD)')) lhs[[i]] = sdvars
```

Comment:
`as.name('names(.SD)'))` AFAIK this is invalid, `as.name` is appropriate for a scalar character name of a variable or function. Here you already do `names` call, so `as.call(list(as.name("names"), as.name(".SD")))`

Reply:
I was surprised that ```as.name('names(.SD)')``` worked, mainly because ```is.name(quote(names(.SD)))``` is *FALSE* but somehow ```quote(names(.SD)) == as.name('names(.SD)')``` is *TRUE*.

I looked into ```quote(names(.SD))``` as well. Using both the ```quote(...)``` method and the ```as.call(list(...))``` methods result in warnings on tests. Using ```as.name(...)``` resulted in no errors on my home PC.
 
Here's an example of one warning:
```
Running test id 352.1      Test 352.1 produced 1 warnings but expected 0
Expected: 
Observed: longer object length is not a multiple of shorter object length
###test 352.1 for reference
test(352.1, DT[,letters[1:4]:=list(1L,NULL)], error="Supplied 4 columns to be assigned 2 items. Please see NEWS for v1.12.2")
```
I also just worked out a recursive function similar to ```replace_dot_alias``` that does not produce errors or warnings and seems more appropriate. That would turn the PR into:
```
if (is.name(lhs)) {
  if (lhs == as.name('.SD')) lhs = sdvars else lhs = as.character(lhs)
} else {
  replace_names_sd = function(e){
    if (length(e) == 1L) return(e)
    if(e[[1L]] == as.name('names') && e[[2L]] == as.name('.SD')) return(sdvars)
    for (i in seq_along(e)[-1L]) if (!is.null(e[[i]])) e[[i]] = replace_names_sd(e[[i]])
    e
  }
  lhs = eval(replace_names_sd(lhs), parent.frame(), parent.frame())
}
```

Reply:
AFAIU using `as.name('names(.SD)')` will be more or less like doing `data.table(`as.name('names(.SD)')` = 1:10)`, this has to be call rather than symbol, which is a product of `as.name`.

Reply:
I made additional changes that should be more robust by using a recursive function to check all calls within LHS for ```expr[[1L]] == as.name('names') && expr[[2L]]== as.name('SD')```.